spd/clustering/activations.py

```diff
@@ -0,0 +1,218 @@
+
+    activations_: dict[str, Float[Tensor, " n_steps C"]]
+    if seq_mode == "concat":
+        # Concatenate the sequence dimension into the sample dimension
+        activations_ = {
+            key: act.reshape(act.shape[0] * act.shape[1], act.shape[2])
+            for key, act in activations.items()
+        }
+    elif seq_mode == "seq_mean":
+        # Take the mean over the sequence dimension
+        activations_ = {
+            key: act.mean(dim=1) if act.ndim == 3 else act for key, act in activations.items()
+        }
+    else:
+        # Use the activations as they are
+        activations_ = activations
+
+    # filter activations for only the modules we want
+    if filter_modules is not None:
+        activations_ = {key: act for key, act in activations_.items() if filter_modules(key)}
+
+    # Sort components within each module if requested
+    sort_indices_dict: dict[str, Int[Tensor, " C"]] = {}
+    if sort_components:
+        sorted_activations = {}
+        for key, act in activations_.items():
+            sorted_act, sort_idx = sort_module_components_by_similarity(act)
+            sorted_activations[key] = sorted_act
+            sort_indices_dict[key] = sort_idx
+        activations_ = sorted_activations
+
+    # compute the labels and total component count
+    total_c: int = 0
+    labels: list[str] = list()
+    for key, act in activations_.items():
+        c = act.shape[-1]
+        if sort_components and key in sort_indices_dict:
+            # Use sorted indices for labeling
+            sort_idx = sort_indices_dict[key]
+            labels.extend([f"{key}:{int(sort_idx[i].item())}" for i in range(c)])
+        else:
+            labels.extend([f"{key}:{i}" for i in range(c)])
+        total_c += c
+
+    # concat the activations
+    act_concat: Float[Tensor, " n_steps c"] = torch.cat(
+        [activations_[key] for key in activations_], dim=-1
+    )
+
+    # filter dead components
+    dead_components_lst: list[str] | None = None
+    if filter_dead_threshold > 0:
+        dead_components_lst = list()
+        max_act = act_concat.max(dim=0).values
+        dead_components = max_act < filter_dead_threshold
+        if dead_components.any():
+            act_concat = act_concat[:, ~dead_components]
+            alive_labels: list[tuple[str, bool]] = [
+                (lbl, bool(keep.item()))
+                for lbl, keep in zip(labels, ~dead_components, strict=False)
+            ]
+            labels = [label for label, keep in alive_labels if keep]
+            dead_components_lst = [label for label, keep in alive_labels if not keep]
+            # logger.values({
+            #     "total_components": total_c,
+            #     "n_alive_components": len(labels),
+            #     "n_dead_components": len(dead_components_lst),
+            # })
+
+    # compute coactivations
+    # TODO: this is wrong for anything but boolean activations
+    coact: Float[Tensor, " c c"] = act_concat.T @ act_concat
```

Comment:
But they're not boolean in our current setup. Also, why is it invalid for non-booleans?